{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ANNOVAR FILES\n",
    "\n",
    "Este programa tiene como objetivo tomar un archivo de OMIM que contiene las variantes mendelianas y regresar un archivo con extensión .avinput el cual pueda ser leido por el programa ANNOVAR. Posteriormente se obtendra la ubicación correspontiente a cada variante y podra graficarse los resultados.\n",
    "\n",
    "### Flujo de los programas\n",
    "\n",
    "Para generar el archivo .avinput sera necesaria la ejecución de dos códigos:\n",
    "\n",
    "1. **getVariants** -> El cual se encarga de obtener el id de las variantes asociadas a un MIMnumber, a partir del archivo genemap2.txt proporcionado pòr OMIM.\n",
    "\n",
    "2. **GenAvinput** -> El cual toma el archivo generado pot getVariants y genera el archivo .avinput que se le suministrara a ANNOVAR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# getVariants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Copyright (c) 1966-2021 Johns Hopkins University. Use of this file adheres to the terms specified at https://omim.org/help/agreement.\n",
      "# Generated: 2021-04-22\n",
      "# See end of file for additional documentation on specific fields\n",
      "# Chromosome\tGenomic Position Start\tGenomic Position End\tCyto Location\tComputed Cyto Location\tMIM Number\tGene Symbols\tGene Name\tApproved Symbol\tEntrez Gene ID\tEnsembl Gene ID\tComments\tPhenotypes\tMouse Gene Symbol/ID\n",
      "chr1\t0\t27600000\t1p36\t\t607413\tAD7CNTP\tAlzheimer disease neuronal thread protein\t\t\t\t\t\t\n",
      "chr1\t0\t27600000\t1p36\t\t612367\tALPQTL2\tAlkaline phosphatase, plasma level of, QTL 2\t\t100196914\t\tlinkage with rs1780324\t{Alkaline phosphatase, plasma level of, QTL 2}, 612367 (2)\t\n",
      "chr1\t0\t123400000\t1p\t\t606788\tANON1\tAnorexia nervosa, susceptibility to, 1\t\t171514\t\t\t{Anorexia nervosa, susceptibility to, 1}, 606788 (2)\t\n",
      "chr1\t0\t27600000\t1p36\t\t605462\tBCC1\tBasal cell carcinoma, susceptibility to, 1\t\t100307118\t\tassociated with rs7538876\t{Basal cell carcinoma, susceptibility to, 1}, 605462 (2)\t\n",
      "chr1\t0\t27600000\t1p36\t\t606928\tBMND3\tBone mineral density QTL 3\t\t246259\t\t?another locus at 3p21\t[Bone mineral density QTL 3], 606928 (2)\t\n",
      "chr1\t0\t2300000\t1p36.33\t\t618815\tC1DUPp36.33, DUP1p36.33\tChromosome 1p36.33 duplication syndrome, ATAD3 gene cluster\t\t\t\t\tChromosome 1p36.33 duplication syndrome, ATAD3 gene cluster, 618815 (4), Autosomal dominant\t\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head ./OMIM/genemap2.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(description='This program creates a file with the monogenic variants in VCF format, which can be used by ANNOVAR to get their genomic location.')\n",
    "parser.add_argument(\n",
    "    \"--input\",\n",
    "    metavar = \"path/to/file\",\n",
    "    help = \"Path to OMIM Gene Map File\",\n",
    "    required = True,\n",
    "    \"--varInfo\",\n",
    "    metavar = \"path/to/file\",\n",
    "    help = \"File path where the user will save the variants information\",\n",
    "    required = True,\n",
    "    \"--VCF\",\n",
    "    metavar = \"path/to/file\",\n",
    "    help = \"File path with monogenic variants in VCF format\",\n",
    "    required = True   \n",
    ")\n",
    "args = parser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################################################################################\n",
    "#####################    Program for the Construction of monogenic variant files     #############################\n",
    "##########################   These files are received as input by ANNOVAR   ######################################\n",
    "##################################################################################################################\n",
    "\n",
    "\n",
    "\n",
    "####################################################################################################################\n",
    "####################################################################################################################\n",
    "                    ############################## Functions ##############################\n",
    "####################################################################################################################\n",
    "####################################################################################################################\n",
    "\n",
    "\n",
    "\n",
    "####################################################################################################################\n",
    "                    ############################## OMIM ##############################\n",
    "####################################################################################################################\n",
    "\n",
    "def get_inheritance(Info,GeneType):\n",
    "    '''\n",
    "    **Description** -> This fucntion return the location of Mendelian diseases\n",
    "    ::Info::(param): String that conteins all the monogenic variant info\n",
    "    ::GeneType::(param): Dictionary that conteins all the monogenic disease inheritance keys\n",
    "    '''\n",
    "    \n",
    "    # Search into the variant to find out if it is a Mendelian disease\n",
    "    Data = None\n",
    "    for key in GeneType:\n",
    "        if key in Info:\n",
    "            \n",
    "            Info = Info.split(\"\\t\")\n",
    "            \n",
    "            # Get interest fields\n",
    "            Chrom = Info[0]\n",
    "            Start = Info[1]\n",
    "            End = Info[2]\n",
    "            MIMnumber = Info[5]\n",
    "            \n",
    "            # Return interest fields in a single tupple\n",
    "            Data = (key,Chrom,Start,End,MIMnumber)\n",
    "            return(Data)\n",
    "    \n",
    "    # If the variant isn't a Mendelian disease return Null\n",
    "    return(Data)\n",
    "\n",
    "####################################################################################################################\n",
    "                    ############################## Ensembl ##############################\n",
    "####################################################################################################################\n",
    "\n",
    "\n",
    "def get_variant(MonogenicDisease,FilePath):\n",
    "    '''\n",
    "    **Description** -> This fucntion find out the variant id of a monogenic disease and then create a file that contains the info of the monogenic disease\n",
    "    ::MonogenicDisease::(param): A dictionary that conteins all the Mendelian diseases locations\n",
    "    ::FilePath::(param): The file path where the user will save the variant Info\n",
    "    '''\n",
    "    c = 0\n",
    "    Header = \"#ID\\tChromosome\\tStart\\tEnd\\tMIM\\tInheritance\\n\"\n",
    "    NewFile = open(FilePath,\"w\")\n",
    "    NewFile.write(Header)\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    \n",
    "    # iterate the dictionary to get the keys of the disease type\n",
    "    for DiseaseType in MonogenicDisease:\n",
    "        for Disease in MonogenicDisease[DiseaseType]:\n",
    "            \n",
    "            # Get the Mendelian disease features\n",
    "            Chromosome = Disease[0].replace(\"chr\",\"\")\n",
    "            Start = Disease[1]\n",
    "            End = Disease[2]\n",
    "            MIMnumber = Disease[3]\n",
    "            \n",
    "            # Search all the variants asociated to this chromosome location\n",
    "            ext = \"/phenotype/region/homo_sapiens/{}:{}-{}?feature_type=Variation;non_specified=0;trait=1;tumour=0\".format(Chromosome,Start,End)\n",
    "            r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "            \n",
    "            decoded = r.json()\n",
    "            for variant in decoded:\n",
    "                try:\n",
    "                    # If the variant contain a MIMnumber we have to compare if it is the MIMnumber that we was searching\n",
    "                    MIM2 = variant[\"phenotype_associations\"][0][\"attributes\"][\"MIM\"]\n",
    "                    ID = variant[\"id\"]\n",
    "                    if MIMnumber == MIM2:\n",
    "                        \n",
    "                        # when we find the MIMnumber of interest we write in the file the variant info\n",
    "                        data = (ID,Chromosome,Start,End,MIMnumber)\n",
    "                        Location = variant[\"phenotype_associations\"][0][\"location\"]\n",
    "                        Location = Location.replace(\":\",\",\").replace(\"-\",\",\")\n",
    "                        Location = Location.split(\",\")\n",
    "                        Start2 =  Location[1]\n",
    "                        End2 = Location[2]\n",
    "                        \n",
    "                        Line = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(ID,Chromosome,Start2,End2,MIMnumber,DiseaseType)\n",
    "                        NewFile.write(Line)\n",
    "                        print(c)\n",
    "                        c = c + 1\n",
    "                        \n",
    "                except:\n",
    "                    continue\n",
    "    NewFile.close()             \n",
    "    return(FilePath)\n",
    "\n",
    "####################################################################################################################\n",
    "####################################################################################################################\n",
    "                    ############################## Main Code ##############################\n",
    "####################################################################################################################\n",
    "####################################################################################################################\n",
    "\n",
    "import requests, sys\n",
    "import re\n",
    "\n",
    "Path = \"./genemap2.txt\"\n",
    "Monogenic = {  \n",
    "            \"Multifactorial\":[],\n",
    "            \"Isolated cases\":[], \n",
    "            \"Digenetic recessive\":[],\n",
    "            \"Somatic mutation\":[],\n",
    "            \"Mitochondrial\":[],\n",
    "            \"Inherited chromosomal imbalance\":[],\n",
    "            \"X-linked dominant\":[],\n",
    "            \"X-linked recessive\":[],\n",
    "            \"Pseudoautosomal recessive\":[],\n",
    "            \"Pseudoautosomal dominant\":[],\n",
    "            \"Autosomal dominant\":[],\n",
    "            \"Autosomal recessive\":[],\n",
    "            \"X-linked\":[]}\n",
    "\n",
    "\n",
    "# Open genemap2 file\n",
    "File = open(Path,\"r\")\n",
    "for line in File:\n",
    "    \n",
    "    # Ignore commented lines\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    \n",
    "    # Eliminate line break and find out info by calling de function\n",
    "    line = line.rstrip()\n",
    "    values = get_inheritance(line,Monogenic)\n",
    "    \n",
    "    if values:\n",
    "        Monogenic[values[0]].append(values[1:])\n",
    "        \n",
    "VariantPath = \"./VariantFileOMIM.txt\"    \n",
    "get_variant(Monogenic,VariantPath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#ID\tChromosome\tStart\tEnd\tMIM\tInheritance\n",
      "rs1569231897\tX\t10213710\t10213710\t302910\tX-linked dominant\n",
      "rs1569233549\tX\t10220876\t10220876\t302910\tX-linked dominant\n",
      "rs1569226551\tX\t10187602\t10187602\t302910\tX-linked dominant\n",
      "rs1569230006\tX\t10206463\t10206463\t302910\tX-linked dominant\n",
      "rs121917889\tX\t11121652\t11121652\t300056\tX-linked dominant\n",
      "rs121917888\tX\t11120974\t11120974\t300056\tX-linked dominant\n",
      "rs387906488\tX\t11294802\t11294810\t300391\tX-linked dominant\n",
      "rs104894737\tX\t11294790\t11294790\t300391\tX-linked dominant\n",
      "rs387906491\tX\t11298779\t11298781\t300391\tX-linked dominant\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head ../downloads/VariantFileOMIM3.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GenAvinput"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#ID\tChromosome\tStart\tEnd\tMIM\tInheritance\n",
      "rs1569231897\tX\t10213710\t10213710\t302910\tX-linked dominant\n",
      "rs1569233549\tX\t10220876\t10220876\t302910\tX-linked dominant\n",
      "rs1569226551\tX\t10187602\t10187602\t302910\tX-linked dominant\n",
      "rs1569230006\tX\t10206463\t10206463\t302910\tX-linked dominant\n",
      "rs121917889\tX\t11121652\t11121652\t300056\tX-linked dominant\n",
      "rs121917888\tX\t11120974\t11120974\t300056\tX-linked dominant\n",
      "rs387906488\tX\t11294802\t11294810\t300391\tX-linked dominant\n",
      "rs104894737\tX\t11294790\t11294790\t300391\tX-linked dominant\n",
      "rs387906491\tX\t11298779\t11298781\t300391\tX-linked dominant\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head ../downloads/VariantFileOMIM3.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "import re\n",
    "\n",
    "# Path to the input and output file\n",
    "Path = \"./VariantFileOMIM3.txt\"\n",
    "ANNOVAR = \"./Monogenic2.avinput\"\n",
    "\n",
    "# Open files\n",
    "File = open(Path,\"r\")\n",
    "avinputFile = open(ANNOVAR,\"w\")\n",
    "Contador = 0\n",
    "\n",
    "for line in File:\n",
    "    \n",
    "    # Ignore first line of the document\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    \n",
    "    # Get file fields\n",
    "    line = line.rstrip()\n",
    "    line = line.split(\"\\t\")\n",
    "    \n",
    "    # Get variant Id\n",
    "    Id = line[0]\n",
    "    Id = Id.replace(\"rs\",\"\")\n",
    "    \n",
    "    # Get variant features\n",
    "    Chromosome = line[1]\n",
    "    Start = line[2]\n",
    "    End = line[3]\n",
    "    MIM = line[4]\n",
    "    DiseaseType = line[5]\n",
    "\n",
    "    # Get the nucleotide change of the variant from NCBI with urllib\n",
    "    UrlPath = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=snp&id={}&retmode=xml'.format(Id)\n",
    "    with urllib.request.urlopen(UrlPath) as response:\n",
    "        html = str(response.read())\n",
    "        \n",
    "        # Search the nucleotide change of the variant\n",
    "        SNP = re.findall('SEQ=\\[.+\\]', html)\n",
    "        Significance = re.findall('<CLINICAL_SIGNIFICANCE>.+</CLINICAL_SIGNIFICANCE>', html)\n",
    "        \n",
    "        # Save the nucleotide change and the clinvar significance\n",
    "        try:\n",
    "            SNPVariant = SNP[0].replace(\"SEQ=[\",\"\").replace(\"]\",\"\")\n",
    "            ClinvarSig = Significance[0].replace(\"<CLINICAL_SIGNIFICANCE>\",\"\").replace(\"</CLINICAL_SIGNIFICANCE>\",\"\")\n",
    "            \n",
    "        # If there is not nucleotide change or clinvar significance go to the next variant\n",
    "        except:\n",
    "            continue\n",
    "        \n",
    "        # Length of the nucleotide change tell us if it is a mutation, insertion or deletion\n",
    "        Blength = len(SNPVariant)\n",
    "        print(Contador)\n",
    "        Contador = Contador + 1\n",
    "        SNPVariant = SNPVariant.split(\"/\")\n",
    "        Alength = len(SNPVariant)\n",
    "        \n",
    "        # If Alength is greater than 2 it means that there is more than 1 polimorfism of th same variant\n",
    "        if Alength > 2:\n",
    "            for i in range(Alength):\n",
    "                \n",
    "                if i == 0:\n",
    "                    continue\n",
    "                \n",
    "                # If the nucleotide start position is a middle dash, the variant is an insertion\n",
    "                if SNPVariant[0] == \"-\":\n",
    "                    Info = \"{} {} {} {} {} comments: rs{}, Insertion, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[i],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                    avinputFile.write(Info)\n",
    "                    continue\n",
    "                \n",
    "                # If the nucleotide change position is a middle dash, the variant is a deletion\n",
    "                if SNPVariant[i] == \"-\":\n",
    "                    Info = \"{} {} {} {} {} comments: rs{}, Deletion, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[i],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                    avinputFile.write(Info)\n",
    "                    continue\n",
    "                \n",
    "                # If the change position is greater than start position, it means that the variant is an insertion\n",
    "                if len(SNPVariant[0]) < len(SNPVariant[i]):\n",
    "                    Info = \"{} {} {} {} {} comments: rs{}, Insertion, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[i],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                    avinputFile.write(Info)\n",
    "                    continue\n",
    "                \n",
    "                # If the start position is greater than change position, it means that the variant is a deletion\n",
    "                if len(SNPVariant[0]) > len(SNPVariant[i]):\n",
    "                    Info = \"{} {} {} {} {} comments: rs{}, Deletion, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[i],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                    avinputFile.write(Info)\n",
    "                    continue\n",
    "                    \n",
    "                # If there is no difference in length after nucleotide change, the variant is a mutation  \n",
    "                Info = \"{} {} {} {} {} comments: rs{}, Mutation, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[i],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                avinputFile.write(Info)\n",
    "                continue\n",
    "                \n",
    "        # If Alength is equal to 1 it means that there is a single polimorifism of the variant       \n",
    "        else:\n",
    "            \n",
    "            # If the nucleotide start position is a middle dash, the variant is an insertion\n",
    "            if SNPVariant[0] == \"-\":\n",
    "                Info = \"{} {} {} {} {} comments: rs{}, Insertion, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[1],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                avinputFile.write(Info)\n",
    "                continue    \n",
    "            \n",
    "            # If the nucleotide change position is a middle dash, the variant is a deletion\n",
    "            if SNPVariant[1] == \"-\":\n",
    "                Info = \"{} {} {} {} {} comments: rs{}, Deletion, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[1],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                avinputFile.write(Info)\n",
    "                continue\n",
    "            \n",
    "            # If the change position is greater than start position, it means that the variant is an insertion\n",
    "            if len(SNPVariant[0]) < len(SNPVariant[1]):\n",
    "                Info = \"{} {} {} {} {} comments: rs{}, Insertion, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[1],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                avinputFile.write(Info)\n",
    "                continue\n",
    "            #If the start position is greater than change position, it means that the variant is a deletion\n",
    "            if len(SNPVariant[0]) > len(SNPVariant[1]):\n",
    "                Info = \"{} {} {} {} {} comments: rs{}, Deletion, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[1],Id,DiseaseType,MIM,ClinvarSig)\n",
    "                avinputFile.write(Info)\n",
    "                continue  \n",
    "              \n",
    "            # If there is no difference in length after nucleotide change, the variant is a mutation\n",
    "            Info = \"{} {} {} {} {} comments: rs{}, Mutation, {}, {}, {}\\n\".format(Chromosome,Start,End,SNPVariant[0],SNPVariant[1],Id,DiseaseType,MIM,ClinvarSig)\n",
    "            avinputFile.write(Info)\n",
    "            continue \n",
    "            \n",
    "# Close files     \n",
    "avinputFile.close()\n",
    "File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X 10213710 10213710 G A comments: rs1569231897, Mutation, X-linked dominant, 302910, pathogenic\n",
      "X 10220876 10220876 G C comments: rs1569233549, Mutation, X-linked dominant, 302910, pathogenic\n",
      "X 10187602 10187602 G A comments: rs1569226551, Mutation, X-linked dominant, 302910, pathogenic\n",
      "X 10206463 10206463 C G comments: rs1569230006, Mutation, X-linked dominant, 302910, pathogenic\n",
      "X 11121652 11121652 C T comments: rs121917889, Mutation, X-linked dominant, 300056, pathogenic\n",
      "X 11120974 11120974 C T comments: rs121917888, Mutation, X-linked dominant, 300056, pathogenic\n",
      "X 11294802 11294810 TTTTATTTG - comments: rs387906488, Deletion, X-linked dominant, 300391, pathogenic\n",
      "X 11294790 11294790 T C comments: rs104894737, Mutation, X-linked dominant, 300391, pathogenic\n",
      "X 11298779 11298781 C - comments: rs387906491, Deletion, X-linked dominant, 300391, pathogenic\n",
      "X 11298833 11298834 C - comments: rs387906489, Deletion, X-linked dominant, 300391, pathogenic\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head ../../downloads/Monogenic2.avinput"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
