---
title: "Plots variants"
author: "Athena Tamayo Luisce"
date: "23/5/2021"
output: html_document
---

```{r}
library(ggplot2)
library(dplyr)
```
Files with variants associated with monogenic diseases 
```{r}
setwd("/home/athena/Escritorio/CCG/4toSemestre/GenoHumana/Proyecto")
data1 <- read.table("PathogenicVariants.variant_function", sep = "\t", header = FALSE)
data2 <- read.table("PathogenicVariantsInvalid.variant_function", sep = "\t", header = FALSE)
exons1 <- read.table("ExonicPathogenicVariants.variant_function", sep = "\t", header = FALSE)
exons2 <- read.table("ExonicPathogenicVariantsInvalid.variant_function", sep = "\t", header = FALSE)
```

Merge data1/data2 and exon1/exon2
```{r}
data <- left_join(data1, data2)
exons <- left_join(exons1, exons2)
```

```{r}
colnames(data) <- c("location", "gene", "description")
head(data, 20)
```

```{r}
dim(data)
dim(exons)
```

Files with the loci associated with monogenic diseases 
```{r}
setwd("/home/athena/Escritorio/CCG/4toSemestre/GenoHumana/Proyecto")
data1 <- read.table("./ANNOVAR_hg38_filtered/Monogenic3.variant_function", sep = "\t", header = FALSE)
data2 <- read.table("./ANNOVAR_hg38_invalid_filtered/Monogenic3_1.variant_function", sep = "\t", header = FALSE)
```

Merge data1/data2 and exon1/exon2
```{r}
loci <- left_join(data1, data2)
colnames(loci) <- c("location", "gene", "description")
```

```{r}
dim(loci)
nrowLoci <- nrow(loci)
```

Plot the frequency of non coding and coding regions with variants and loci associated with monogenic diseases
```{r}
# create dataframe of variant frequencies 
freq <- as.data.frame(table(data$location))
# add the percentages 
freq$Perc <- round((freq$Freq)/sum(freq$Freq)*100, 2)

# create dataframe of the loci frequencies 
freqLoci <- as.data.frame(table(loci$location))
# add the percentages 
freqLoci$Perc <- round((freqLoci$Freq)/sum(freqLoci$Freq)*100, 2)
```

```{r}
# Get the frequency of the coding and non-coding regions with variants associated with monogenic
regions1 <- data.frame(region = c("coding", "non-coding"), 
                      freq = c(sum(freq$Freq[c(1, 2)]), 
                               sum(freq$Freq[c(3, 4, 5, 6, 7, 8, 9, 10, 11)])),
                      genomic = c("Variants", "Variants"))
regions1$perc <- round(regions1$freq/sum(freq$Freq)*100, 2)

# Get the frequency of the coding and non-coding regions with loci associated with monogenic
regions2 <- data.frame(region = c("coding", "non-coding"), 
                      freq = c(sum(freqLoci$Freq[c(1, 2)]), 
                               sum(freqLoci$Freq[c(3, 4, 5, 6, 7, 8, 9, 10, 11)])),
                      genomic = c("Loci", "Loci"))
regions2$perc <- round(regions2$freq/sum(freqLoci$Freq)*100, 2)
```

```{r}
# Merge both dataframes 
regions <- rbind(regions1, regions2)
```

```{r}
ggplot(regions, aes(x = genomic, y = freq, fill = region)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme(plot.title = element_text(face = "bold", hjust = 0.05),
        legend.title =  element_text(face = "bold")) +
  labs(title = "Frequency of variants and loci associated with monogenic diseases
                     located in coding and non-coding regions", x = NULL, y = "Frequency") +
  scale_fill_discrete(name = "Genomic region") +
  geom_text(aes(label = paste(perc, "%", sep = "")), vjust = -0.2, position = position_dodge(width = 0.9))
```

Plot the frequency of mutations associated with monogenic diseases located in different genomic regions.
```{r}
ggplot(freq, aes(x = Var1, y = log(Freq), fill = Var1)) +
  geom_bar(stat = "identity") +
  theme(axis.ticks.x = element_blank(),
        axis.text.x=element_blank(),
        plot.title = element_text(face = "bold", hjust = 0.05),
        legend.title =  element_text(face = "bold")) +
  labs(title = "Frequency of variants associated with monogenic diseases located 
                            in different genomic regions", x = NULL, y = "log(frequency)") +
  scale_fill_discrete(name = "Genomic location") +
  geom_text(aes(label = paste(Perc, "%", sep = "")), vjust = -0.2) 
```
Label info: https://annovar.openbioinformatics.org/en/latest/user-guide/gene/

Plot the frequency of the types of the type of mutations associated with monogenic diseases in exons 
```{r}
head(exons)
```
```{r}
mut <- as.data.frame(table(exons$V2))
mut$Perc <- round((mut$Freq)/sum(mut$Freq)*100, 2)
```

```{r}
ggplot(mut, aes(x = Var1, y = log(Freq), fill = Var1)) +
  geom_bar(stat = "identity") +
  theme(axis.ticks.x = element_blank(),
        axis.text.x=element_blank(),
        plot.title = element_text(face = "bold", hjust = 0.05),
        legend.title =  element_text(face = "bold")) +
  labs(title = "Frequency of the type of mutations that gave rise to the exonic variants",
       x = NULL, y = "log(frequency)") +
  scale_fill_discrete(name = "Type of mutations") +
  geom_text(aes(label = paste(Perc, "%", sep = "")), vjust = -0.2) +
  ylim(0, 10)
```

Split the description of the variants and loci 
```{r}
elements <- strsplit(data$description, split = "[,]")
elemLoci <- strsplit(loci$description, split = "[,]")
```

Plot the type of inheritance of the variants and loci 
```{r}
inheritance <- vector()
for (i in 1:15633){
  inheritance <- c(inheritance, elements[[i]][3])
}

inheritance <- as.data.frame(table(inheritance))
inheritance$Perc <- round((inheritance$Freq)/sum(inheritance$Freq)*100, 2)
inheritance$Genomic <- rep("Variants", 6)
```

```{r}
inhLoci <- vector()
for (i in 1:nrowLoci){
  inhLoci <- c(inhLoci, elemLoci[[i]][3])
}

inhLoci <- as.data.frame(table(inhLoci))
inhLoci$Perc <- round((inhLoci$Freq)/sum(inhLoci$Freq)*100, 2)
inhLoci$Genomic <- rep("Loci", 6)
colnames(inhLoci)[1] <- "inheritance"
```

```{r}
inheritance <- rbind(inheritance, inhLoci)
```

```{r}
ggplot(inheritance, aes(x = Genomic, y = log(Freq), fill = inheritance)) +
  geom_bar(stat = "identity", position=position_dodge()) +
  theme(plot.title = element_text(face = "bold", hjust = 0.05),
        legend.title =  element_text(face = "bold")) +
  labs(title = "Frequency of the type of inheritance of the diseases associated \n                            with the variants and loci",
       x = NULL, y = "log(frequency)") +
  scale_fill_discrete(name = "Type of inheritance") +
  geom_text(aes(label = paste(Perc, "%", sep = "")), vjust = -0.2, 
            position = position_dodge(width = 0.9)) +
  ylim(0, 10)
```

Plot the number of diseases associated with a number of variants
```{r}
disease <- vector()
for (i in 1:15633){
  disease <- c(disease, elements[[i]][4])
}
```

```{r}
disease <- as.data.frame(table(disease))
```

```{r}
summary(disease$Freq)
```
```{r}
table(disease$Freq)
```

```{r}
ggplot(disease, aes(x = Freq)) + 
  geom_histogram(binwidth = 1, color = "black", fill = "darkseagreen1") +
  labs(title = "Number of genes with a certain number of variants associated\nwith monogenic diseases",
       x = "Number of variants", y = "Number of genes") +
  theme(plot.title = element_text(face = "bold")) +
  scale_x_continuous(breaks = seq(1, 32, 2), limits = c(0, 30))
```

```{r}
# Obtain the mim number of the disease that is associated with 166 variants 
print(which(disease$Freq == 173))
disease$disease[796]
```
Plot the mutations that give rise to variants and loci associated with monogenic diseases in all the genomic region

Get the mutations that give rice to the loci associated with the monogenic diseases 

```{r}
# Get the mutations associated with monogenic diseases in all the genomic regions 
mutLoci <- vector()
for (i in 1:nrowLoci){
  mutLoci <- c(mutLoci, elemLoci[[i]][2])
}
```

```{r}
mutLoci <- as.data.frame(table(mutLoci))
colnames(mutLoci)[1] <-"mutations"
mutLoci$Perc <- round((mutLoci$Freq)/sum(mutLoci$Freq)*100, 2)
mutLoci$Genomic <- rep("Loci", 3)
```

Get the mutations that give rice to the variants associated with the monogenic diseases 
```{r}
# Get the mutations associated with monogenic diseases in all the genomic regions 
mutations <- vector()
for (i in 1:15633){
  mutations <- c(mutations, elements[[i]][2])
}
```

```{r}
mutations <- as.data.frame(table(mutations))
mutations$Perc <- round((mutations$Freq)/sum(mutations$Freq)*100, 2)
mutations$Genomic <- rep("Variants", 3)
```

Merge both dataframes 
```{r}
mutations <- rbind(mutations, mutLoci)
```

```{r}
ggplot(mutations, aes(x = Genomic, y = Freq, fill = mutations)) +
  geom_bar(stat = "identity", position=position_dodge()) +
  theme(plot.title = element_text(face = "bold", hjust = 0.05),
        legend.title =  element_text(face = "bold")) +
  labs(title = "Frequency of the type of mutations associated with monogenic 
                     diseases in all the genomic regions",
       x = NULL, y = "Frequency") +
  scale_fill_discrete(name = "Type of mutations") +
  geom_text(aes(label = paste(Perc, "%", sep = "")), vjust = -0.2, 
            position = position_dodge(width = 0.9))
```

Plot the distribution of the mutations on each chromosomes
```{r}
desc <- strsplit(data$description, split = "[ ]")
chrs <- vector()
for (i in 1:15633){
   chrs <- c(chrs, desc[[i]][1])
}

chrs <- as.data.frame(table(chrs))
order <- order(c(1,10,11,12,13,14,15,16,17,18,19,2,20,21,22,3,4,5,6,7,8,9,23,24))
```

```{r}
chrs <- data.frame(chrs = chrs$chrs[order], Freq = chrs$Freq[order])
chrs$chrs <- factor(chrs$chrs, unique(chrs$chrs))
```

```{r}
palette <- grDevices::colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]
colors <- sample(palette, 24, replace = FALSE)

ggplot(chrs, aes(x = "", y = Freq, fill = chrs)) +
  geom_bar(stat = "identity", color='black') +
  coord_polar("y", start = 0) +
  theme(axis.text = element_blank(), 
        axis.ticks = element_blank(), 
        panel.grid  = element_blank(),
        plot.title = element_text(face = "bold"),
        legend.title =  element_text(face = "bold"))+
  scale_fill_discrete(name = "Chromosomes", type = colors) +
  labs(title = "Distribution of the mutations on each chromosome", x = "", y = "") 
```

```{r}
chrs$Perc <- round(chrs$Freq/sum(chrs$Freq)*100, 2)
```

```{r}
ggplot(chrs, aes(x = chrs, y = Freq, fill = chrs)) +
  geom_bar(stat = "identity") +
  theme(axis.ticks.x = element_blank(),
        axis.text.x=element_blank(),
        plot.title = element_text(face = "bold", hjust = 0.05),
        legend.title =  element_text(face = "bold")) +
  labs(title = "Frequency of variants in each chromosome",
       x = NULL, y = "Frequency") +
  scale_fill_discrete(name = "Chromosome") +
  geom_text(aes(label = paste(Perc, "%", sep = "")), vjust = -0.2) +
  ylim(c(0, 2000))
```

Chromosome sizes 
```{r}
size <- data.frame(chrs = c(1:22, "X", "Y"), size = c(249, 237, 192, 183, 174, 165, 153, 135, 132, 132, 132, 123, 108, 105, 99, 84, 81, 75, 69, 63, 54, 57, 141, 60)) 
size$chrs <- factor(size$chrs, unique(size$chrs))
```

```{r}
ggplot(size, aes(x = chrs, y = size, fill = chrs)) +
  geom_bar(stat = "identity") +
  theme(axis.ticks.x = element_blank(),
        axis.text.x=element_blank(),
        plot.title = element_text(face = "bold", hjust = 0.05),
        legend.title =  element_text(face = "bold")) +
  labs(title = "Chromosome size (Mbp)",
       x = NULL, y = "Size (Mbp)") +
  scale_fill_discrete(name = "Chromosome") +
  geom_text(aes(label = size), vjust = -0.2) +
  ylim(c(0, 300))
```

Number of genes per chromosome 
```{r}
nGenes <- data.frame(chrs = c(1:22, "X", "Y"), size = c(1963, 1192, 1025, 730, 841, 993, 862, 647, 733, 701, 1225, 981, 310, 586, 558, 793, 1122, 258, 1361, 518, 218, 414, 821, 62)) 
nGenes$chrs <- factor(nGenes$chrs, unique(nGenes$chrs))
```

```{r}
ggplot(nGenes, aes(x = chrs, y = size, fill = chrs)) +
  geom_bar(stat = "identity") +
  theme(axis.ticks.x = element_blank(),
        axis.text.x=element_blank(),
        plot.title = element_text(face = "bold", hjust = 0.05),
        legend.title =  element_text(face = "bold")) +
  labs(title = "Number of genes in every chromosome",
       x = NULL, y = "Number of genes") +
  scale_fill_discrete(name = "Chromosome") +
  geom_text(aes(label = size), vjust = -0.2) +
  ylim(c(0, 2000))
```
